[Monaco] Add llvm ir syntax highlighting

bugwelle opened the pull request:
Hi,

this PR adds basic syntax highlighting for LLVM IR (editor pane).
I used the official [vim syntax highlighting](http://llvm.org/svn/llvm-project/llvm/trunk/utils/vim/syntax/llvm.vim) so hopefully all tokens are recognized correctly :-)

Note: We may want to change some monaco classes in `static/modes/llvm-ir-mode.js`. I chose the ones I saw fit.

PS: I also renamed `llvm.defaults.properties` (the file was named incorrectly).


Regards,
Andre

bugwelle commented on static/modes/llvm-ir-mode.js:
```diff
@@ -0,0 +1,141 @@
+            'type', 'label', 'opaque', 'token'
+        ],
+        // llvmStatement
+        śtatements: [
```
Yes. `llvmStatement` is the name used by the VIM syntax definition. I'd call them keywords as well. GitHub colors keywords and statements the same, e.g.

```llvm
%3 = alloca %"class.std::exception"*, align 8
attributes #6 = { noreturn }
```

I'm open for a different way of handling statements. :-)

bugwelle replied:
Oh, my bad. Thought you meant something else. Yeah, the `s` is a typo.

partouf commented:
Because the default is on invalid, I can spot a couple of things that aren't covered by the styling yet.

I suggest adding at least:
```
[/[,(){}]/, 'delimiters'], <-- add somewhere on top
[/[*#=!]/, 'operators'], <-- add this to the bottom
```

There's a couple of other things broken, like parameter lists 'name: value' that aren't handled properly, but I can't seem to figure out how to solve that.
So I suggest to leave that and just remove default: 'invalid' , unless you know a way.

bugwelle commented:
@partouf Thank you. I had `defaultToken: 'invalid'` commented out but didn't push it. I've removed it again. I've added the missing operators and delimiters (thanks for noticing that).

The LLVM IR highlighting used by GitHub is: https://github.com/whitequark/LLVM.tmBundle/blob/master/Syntaxes/LLVM.tmLanguage
It has fewer tokens and does not handle special blocks but is mostly the same with this version. 

The LLVM IR lexer used by the LLVM IR documentation (pygments) can be found here: https://bitbucket.org/birkenfeld/pygments-main/src/7941677dc77d4f2bf0bbd6140ade85a9454b8b80/pygments/lexers/asm.py?at=default&fileviewer=file-view-default#asm.py-335
They don't distinguish between statements and "normal" keywords. Should we do the same?

Regards,
Andre

RabsRincon commented:
Hi @bugwelle. Sorry about the radio silence. We're holding UI changes until #1105 is merged, as it upgrades a good chunk of the UI. (Thus the lack of activity both here and on the llvm ir pane PRs)

We hope to have it merged upstream this week, so please bear with us while we iron the last few bugs out

:)

bugwelle commented:
Rebased both commits :smiley:

bugwelle commented:
Rebased both commits :smiley:

mattgodbolt commented:
@partouf - now it's been rebased do you have any other concerns or thoughts?

mattgodbolt commented on static/modes/llvm-ir-mode.js:
```diff
@@ -0,0 +1,140 @@
+
+        tokenizer: {
+            root: [
+                [/[,(){}<>\[\]]/, 'delimiters'],
```
The `\` before the `[` is unnecessary (and the IDE I use gives it a wiggly red underline. It seems OK to change this to:

```
                [/[,(){}<>[\]]/, 'delimiters'],
```

bugwelle replied:
Indeed, because we're already inside `[]`.

```suggestion
                [/[,(){}<>[\]]/, 'delimiters'],
```

mattgodbolt replied:
I'd probably keep them: it's expressive and for all I know it might help the editor out in some way (maybe not now, but in future?). We're looking to upgrade Monaco in a separate PR.

partouf commented:
I see no reason not to merge

mattgodbolt commented:
Thanks everyone!
